odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return (self.scalar * self.operator.gradient *
+                ScalingOperator(self.domain, self.scalar))
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of functional with scaled argument."""
+        return ConvexConjugateArgScaling(self.operator.conjugate_functional,
+                                         self.scalar)
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the functional with scaled argument.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+        """
+        sigma = float(sigma)
+        return(proximal_arg_scaling(
+            self.operator.proximal, self.scalar))(sigma)
+
+    def derivative(self, point):
+        """Returns the derivative operator in the given point.
+
+        This function returns the linear operator
+
+            ``x --> <x, grad_f(point)>``,
+
+        where ``grad_f(point)`` is the gradient of the functional in the point
+        ``point``.
+
+        Parameters
+        ----------
+        point : `LinearSpaceVector`
+            The point in which the gradient is evaluated.
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        return self.scalar * self.operator.derivative(self.scalar * point)
+
+
+class FunctionalComp(Functional, OperatorComp):
+
+    """Composition of a functional with an operator.
+
+    Given a functional ``func`` and an operator ``op``, such that the range of
+    the operator is equal to the domain of the functional, this corresponds to
+    the functional
+
+        ``(func * op)(x) = func(op(x))``.
+    """
+
+    def __init__(self, func, op, tmp1=None, tmp2=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            The left ("outer") operator
+        op : `Operator`
+            The right ("inner") operator. Its range must coincide with the
+            domain of ``func``.
+        tmp1 : `element` of the range of ``op``, optional
+            Used to avoid the creation of a temporary when applying ``op``
+        tmp2 : `element` of the range of ``op``, optional
+            Used to avoid the creation of a temporary when applying the
+            gradient of ``func``
+        """
+        if not isinstance(func, Functional):
+            raise TypeError('functional {!r} is not a Functional instance.'
+                            ''.format(func))
+
+        OperatorComp.__init__(self, left=func, right=op, tmp=tmp1)
+
+        # TODO: can the Lipschitz constant of the gradient be estimated?
```

Comment:
I would guess you can estimate `new_lip <= opnorm * old_lip` of the operator is linear, but there is no direct "non-computational" way to get the operator norm other than adding it as an extra parameter for the user to supply. But then, you would need to check if the operator is linear or not and just apply the estimate in this case. Frankly, I don't think this is worth the trouble since this class will very very seldom be used in cases other than `func * op`, which uses default arguments.